timemachine/fe/cif_writer.py

```diff
@@ -0,0 +1,167 @@
+        for obj_idx, obj in enumerate(objs):
+            old_topology = obj
+            if isinstance(obj, app.Topology):
+                old_chain_id_kv = {}
+                for old_chain in old_topology.chains():
+                    new_chain = combined_topology.addChain()
+                    old_chain_id_kv[old_chain.id] = new_chain
+
+                old_atom_id_kv = {}
+                for old_residue in old_topology.residues():
+                    chain_obj = old_chain_id_kv[old_residue.chain.id]
+                    new_residue = combined_topology.addResidue(name=old_residue.name, chain=chain_obj)
+                    for old_atom in old_residue.atoms():
+                        new_atom = combined_topology.addAtom(old_atom.name, old_atom.element, new_residue)
+                        old_atom_id_kv[old_atom.id] = new_atom
+
+                for old_bond in old_topology.bonds():
+                    old_atom1_id = old_bond.atom1.id
+                    old_atom2_id = old_bond.atom2.id
+                    new_atom1 = old_atom_id_kv[old_atom1_id]
+                    new_atom2 = old_atom_id_kv[old_atom2_id]
+                    combined_topology.addBond(new_atom1, new_atom2, old_bond.type, old_bond.order)
+
+            elif isinstance(obj, Chem.Mol):
+                mol = obj
+                new_chain = combined_topology.addChain()
+                new_residue = combined_topology.addResidue(name="LIG", chain=new_chain)
+                old_idx_to_new_atom_map = {}
+                for atom in mol.GetAtoms():
+                    name = atom.GetSymbol() + str(atom.GetIdx())
+                    element = app.element.Element.getByAtomicNumber(atom.GetAtomicNum())
+                    new_atom = combined_topology.addAtom(name, element, new_residue)
+                    old_idx = atom.GetIdx()
+                    old_idx_to_new_atom_map[old_idx] = new_atom
+
+                # (ytz): while we could get fancier, mmcif only has the 'covale' bond type:
+                # https://mmcif.wwpdb.org/dictionaries/mmcif_pdbx_v40.dic/Items/_struct_conn_type.id.html
+                # so we only bother with "yes/no"
+                for bond in mol.GetBonds():
+                    src_atom = old_idx_to_new_atom_map[bond.GetBeginAtomIdx()]
+                    dst_atom = old_idx_to_new_atom_map[bond.GetEndAtomIdx()]
+                    combined_topology.addBond(src_atom, dst_atom)
+
+            else:
+                raise ValueError(f"Unknown obj type: {type(obj)}")
+
+        self.topology = combined_topology
+
+        self.out_handle = open(out_filepath, "w")
+        PDBxFile.writeHeader(self.topology, self.out_handle)
+        self.topology = self.topology
+        self.frame_idx = 0
+
+    def write_frame(self, x):
+        """
+        Write a coordinate frame
+
+        Parameters
+        ----------
+        x: np.ndarray
+            coordinates in units of angstroms
+
+        """
+        self.frame_idx += 1
```

Comment:
nit: To verify, the frame idx is 1 indexed? Not sure if it matters

Reply:
yes - same as PDB, things like `HETATM` records are all 1-indexed